### ISSUE TITLE
[Feature]:  request __shfl_xor_sync  in rocm

### ISSUE THREAD

**Issue by ZJLi2013** [ROCm/ROCm#3917]
hi, rocm expert, 
 during our recent project, we need a same function as cuda, __shfl_xor_sync().

however looks recent [rocm 6.2.2](https://rocm.docs.amd.com/projects/HIPIFY/en/latest/tables/CUDA_Device_API_supported_by_HIP.html) has not supported yet:

|   |     |    |    | 
|---|---|---|---|
__shfl_xor | 9.0 |      __shfl_xor | 1.6.0 | 
__shfl_xor_sync |   |   |   |  

is there other primitives as quick replacement ? 

Thanks
David 

### GPU

mi300

### ROCm Component

6.2.2

**Comment by ZJLi2013** [comment]
chatgpt gave a solution:

```c++
inline int custom_shfl_xor_sync(unsigned mask, int val, int laneMask) {
    // 使用 ROCm intrinsic 来模拟 __shfl_xor_sync 的行为
    // __builtin_amdgcn_permlane_xor 可以在 wavefront 中根据 XOR 掩码交换值
    int result = __builtin_amdgcn_permlane_xor(val, laneMask, 0);
    return result;
}
``` 

thanks for clarifying

**Comment by harkgill-amd** [comment]
Will work on getting the HIPIFY docs updated to remove any confusion. Thanks for pointing this out!

**Comment by ZJLi2013** [comment]
> Hi @ZJLi2013, `__shfl_xor_sync` is supported as of ROCm 6.2, you can find it's definition [here](https://github.com/ROCm/clr/blob/amd-staging/hipamd/include/hip/amd_detail/amd_warp_sync_functions.h#L271-L282).
> 
> For more information regarding it's usage, please see [Warp shuffle functions](https://rocm.docs.amd.com/projects/HIP/en/latest/reference/cpp_language_extensions.html#warp-shuffle-functions). You can also find the example [hipShflSyncXorTests ](https://github.com/ROCm/hip-tests/blob/develop/catch/unit/warp/hipShflSyncXorTests.cc)in our hip-tests repo.

hi, @harkgill-amd , thanks. looks the following header files are enough

```c++
#include <hip/hip_runtime.h>
#include <hip/hip_runtime_api.h>
#include <hip/amd_detail/amd_warp_sync_functions.h>  // __shuffle_xor_sync
```

correct me if I was wrong, thanks

**Comment by fxmarty-amd** [comment]
On ROCm 6.3.0, compiling with `-DHIP_ENABLE_WARP_SYNC_BUILTINS` while using `__half` and `__nv_bfloat16` eventually leads me to

```
In file included from /opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:27,
                 from /opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h:113,
                 from /opt/rocm-6.3.0/include/hip/hip_bf16.h:29,
                 from /quark/quark/torch/kernel/hw_emulation/csrc/mx_fp4/fake_hip.h:6,
                 from /quark/quark/torch/kernel/hw_emulation/csrc/python_function_export_hip.cpp:13:
/opt/rocm-6.3.0/include/hip/amd_detail/device_library_decls.h:71:46: error: ‘_Float16’ does not name a type; did you mean ‘_Float64’?
   71 | extern "C" __device__ __attribute__((const)) _Float16 __ocml_cvtrtn_f16_f32(float);
      |                                              ^~~~~~~~
      |                                              _Float64
/opt/rocm-6.3.0/include/hip/amd_detail/device_library_decls.h:72:46: error: ‘_Float16’ does not name a type; did you mean ‘_Float64’?
   72 | extern "C" __device__ __attribute__((const)) _Float16 __ocml_cvtrtp_f16_f32(float);
      |                                              ^~~~~~~~
      |                                              _Float64
/opt/rocm-6.3.0/include/hip/amd_detail/device_library_decls.h:73:46: error: ‘_Float16’ does not name a type; did you mean ‘_Float64’?
   73 | extern "C" __device__ __attribute__((const)) _Float16 __ocml_cvtrtz_f16_f32(float);
      |                                              ^~~~~~~~
      |                                              _Float64
In file included from /opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h:113,
                 from /opt/rocm-6.3.0/include/hip/hip_bf16.h:29,
                 from /quark/quark/torch/kernel/hw_emulation/csrc/mx_fp4/fake_hip.h:6,
                 from /quark/quark/torch/kernel/hw_emulation/csrc/python_function_export_hip.cpp:13:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h: In function ‘unsigned int __hip_ds_bpermute(int, unsigned int)’:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:38:13: error: ‘__builtin_amdgcn_ds_bpermute’ was not declared in this scope
   38 |     tmp.i = __builtin_amdgcn_ds_bpermute(index, tmp.i);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h: In function ‘float __hip_ds_bpermutef(int, float)’:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:44:13: error: ‘__builtin_amdgcn_ds_bpermute’ was not declared in this scope
   44 |     tmp.i = __builtin_amdgcn_ds_bpermute(index, tmp.i);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h: In function ‘unsigned int __hip_ds_permute(int, unsigned int)’:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:50:13: error: ‘__builtin_amdgcn_ds_permute’ was not declared in this scope
   50 |     tmp.i = __builtin_amdgcn_ds_permute(index, tmp.i);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h: In function ‘float __hip_ds_permutef(int, float)’:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:56:13: error: ‘__builtin_amdgcn_ds_permute’ was not declared in this scope
   56 |     tmp.i = __builtin_amdgcn_ds_permute(index, tmp.i);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h: In function ‘int __hip_move_dpp_N(int)’:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:82:12: error: there are no arguments to ‘__builtin_amdgcn_mov_dpp’ that depend on a template parameter, so a declaration of ‘__builtin_amdgcn_mov_dpp’ must be available [-fpermissive]
   82 |     return __builtin_amdgcn_mov_dpp(src, dpp_ctrl, row_mask, bank_mask,
      |            ^~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:82:12: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h: At global scope:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:87:33: error: ‘__AMDGCN_WAVEFRONT_SIZE’ was not declared in this scope
   87 | static constexpr int warpSize = __AMDGCN_WAVEFRONT_SIZE;
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h: In function ‘long long unsigned int __ballot(int)’:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:108:12: error: ‘__builtin_amdgcn_uicmp’ was not declared in this scope; did you mean ‘__builtin_strncmp’?
  108 |     return __builtin_amdgcn_uicmp(predicate, 0, ICMP_NE);
      |            ^~~~~~~~~~~~~~~~~~~~~~
      |            __builtin_strncmp
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h: In function ‘long long unsigned int __ballot64(int)’:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:114:12: error: ‘__builtin_amdgcn_uicmp’ was not declared in this scope; did you mean ‘__builtin_strncmp’?
  114 |     return __builtin_amdgcn_uicmp(predicate, 0, ICMP_NE);
      |            ^~~~~~~~~~~~~~~~~~~~~~
      |            __builtin_strncmp
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h: In function ‘unsigned int __lane_id()’:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:130:13: error: ‘__builtin_amdgcn_mbcnt_lo’ was not declared in this scope
  130 |         -1, __builtin_amdgcn_mbcnt_lo(-1, 0));
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:129:13: error: ‘__builtin_amdgcn_mbcnt_hi’ was not declared in this scope
  129 |     return  __builtin_amdgcn_mbcnt_hi(
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h: In function ‘int __shfl(int, int, int)’:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:138:12: error: ‘__builtin_amdgcn_ds_bpermute’ was not declared in this scope
  138 |     return __builtin_amdgcn_ds_bpermute(index<<2, var);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h: In function ‘int __shfl_up(int, unsigned int, int)’:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:243:12: error: ‘__builtin_amdgcn_ds_bpermute’ was not declared in this scope
  243 |     return __builtin_amdgcn_ds_bpermute(index<<2, var);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h: In function ‘int __shfl_down(int, unsigned int, int)’:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:349:12: error: ‘__builtin_amdgcn_ds_bpermute’ was not declared in this scope
  349 |     return __builtin_amdgcn_ds_bpermute(index<<2, var);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h: In function ‘int __shfl_xor(int, int, int)’:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_warp_functions.h:452:12: error: ‘__builtin_amdgcn_ds_bpermute’ was not declared in this scope
  452 |     return __builtin_amdgcn_ds_bpermute(index<<2, var);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /opt/rocm-6.3.0/include/hip/hip_bf16.h:29,
                 from /quark/quark/torch/kernel/hw_emulation/csrc/mx_fp4/fake_hip.h:6,
                 from /quark/quark/torch/kernel/hw_emulation/csrc/python_function_export_hip.cpp:13:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h: At global scope:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h:691:74: error: ‘__AMDGCN_WAVEFRONT_SIZE’ was not declared in this scope
  691 |                                                        const int width = __AMDGCN_WAVEFRONT_SIZE) {
      |                                                                          ^~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h:702:75: error: ‘__AMDGCN_WAVEFRONT_SIZE’ was not declared in this scope
  702 |                                                         const int width = __AMDGCN_WAVEFRONT_SIZE) {
      |                                                                           ^~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h: In function ‘__hip_bfloat162 __shfl_down_sync(long long unsigned int, __hip_bfloat162, unsigned int, int)’:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h:708:27: error: expected primary-expression before ‘unsigned’
  708 |   u.ui = __shfl_down_sync<unsigned long long, unsigned int>(mask, u.ui, delta, width);
      |                           ^~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h: At global scope:
/opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h:718:69: error: ‘__AMDGCN_WAVEFRONT_SIZE’ was not declared in this scope
  718 |                                                   const int width = __AMDGCN_WAVEFRONT_SIZE) {
      |                                                                     ^~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h:728:70: error: ‘__AMDGCN_WAVEFRONT_SIZE’ was not declared in this scope
  728 |                                                    const int width = __AMDGCN_WAVEFRONT_SIZE) {
      |                                                                      ^~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h:745:72: error: ‘__AMDGCN_WAVEFRONT_SIZE’ was not declared in this scope
  745 |                                                      const int width = __AMDGCN_WAVEFRONT_SIZE) {
      |                                                                        ^~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h:756:73: error: ‘__AMDGCN_WAVEFRONT_SIZE’ was not declared in this scope
  756 |                                                       const int width = __AMDGCN_WAVEFRONT_SIZE) {
      |                                                                         ^~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h:772:73: error: ‘__AMDGCN_WAVEFRONT_SIZE’ was not declared in this scope
  772 |                                                       const int width = __AMDGCN_WAVEFRONT_SIZE) {
      |                                                                         ^~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-6.3.0/include/hip/amd_detail/amd_hip_bf16.h:782:74: error: ‘__AMDGCN_WAVEFRONT_SIZE’ was not declared in this scope
  782 |                                                        const int width = __AMDGCN_WAVEFRONT_SIZE) {
      |                                                                          ^~~~~~~~~~~~~~~~~~~~~~~
```
